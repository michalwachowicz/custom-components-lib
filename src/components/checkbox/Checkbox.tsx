import { useId } from "react";
import * as styles from "./Checkbox.module.scss";
import Entity from "../clickEntity/ClickEntity";
import Switchable from "../../interfaces/Switchable";
import useFinalRef from "../../hooks/useFinalRef";
import useClickCenterEntity from "../clickEntity/useClickCenterEntity";

export default function Checkbox({
  ref,
  label = undefined,
  checked = false,
  disabled = false,
  onChange,
}: Switchable) {
  const id = useId();
  const finalRef = useFinalRef(ref);

  const { entity, size, spawnEntity } = useClickCenterEntity(finalRef);

  const changeHandler = () => {
    if (onChange) onChange(!checked);
    spawnEntity();
  };

  return (
    <div className={styles["checkbox-container"]}>
      <div className={styles["checkbox-wrapper"]}>
        <input
          ref={finalRef}
          className={styles["checkbox-input"]}
          id={id}
          type='checkbox'
          checked={checked}
          disabled={disabled}
          onChange={changeHandler}
        />

        {entity !== null && (
          <Entity key={entity.id} size={size} entity={entity} />
        )}
      </div>

      {label && (
        <label htmlFor={id} className={styles["checkbox-label"]}>
          {label}
        </label>
      )}
    </div>
  );
}
