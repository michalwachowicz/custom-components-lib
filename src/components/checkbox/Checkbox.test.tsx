import { render, screen } from "@testing-library/react";
import Checkbox from "./Checkbox";
import userEvent from "@testing-library/user-event";
import { act } from "react";

describe("<Checkbox />", () => {
  it("renders with label properly", () => {
    render(<Checkbox label='Test' />);

    expect(screen.getByText("Test")).toBeInTheDocument();
  });

  it("is checked when checked prop is true", () => {
    render(<Checkbox checked />);
    expect(screen.getByRole("checkbox")).toBeChecked();
  });

  it("is not checked when checked prop is false", () => {
    render(<Checkbox />);
    expect(screen.getByRole("checkbox")).not.toBeChecked();
  });

  it("is disabled when disabled prop is true", () => {
    render(<Checkbox disabled />);
    expect(screen.getByRole("checkbox")).toBeDisabled();
  });

  it("is not disabled when disabled prop is false", () => {
    render(<Checkbox />);
    expect(screen.getByRole("checkbox")).not.toBeDisabled();
  });

  it("runs onChange when clicked", async () => {
    const onChange = jest.fn();
    const user = userEvent.setup();

    render(<Checkbox onChange={onChange} />);
    expect(onChange).toHaveBeenCalledTimes(0);

    await act(async () => {
      await user.click(screen.getByRole("checkbox"));
    });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(true);
  });

  it("does not run onChange when disabled", async () => {
    const onChange = jest.fn();
    const user = userEvent.setup();

    render(<Checkbox onChange={onChange} disabled />);

    await act(async () => {
      await user.click(screen.getByRole("checkbox"));
    });

    expect(onChange).toHaveBeenCalledTimes(0);
  });

  it("spawns entity on click", async () => {
    jest.useFakeTimers();

    render(<Checkbox />);
    const user = userEvent.setup({ delay: null });

    await act(async () => {
      await user.click(screen.getByRole("checkbox"));
    });

    expect(screen.getByTestId("entity")).toBeInTheDocument();

    act(() => {
      jest.runAllTimers();
    });

    expect(screen.queryByTestId("entity")).not.toBeInTheDocument();
    jest.useRealTimers();
  });

  it("does not spawn entity when disabled", async () => {
    render(<Checkbox disabled />);

    const user = userEvent.setup();

    await act(async () => {
      await user.click(screen.getByRole("checkbox"));
    });

    expect(screen.queryByTestId("entity")).not.toBeInTheDocument();
  });

  describe("snapshots", () => {
    const labels = [undefined, "Test"];
    const booleans = [true, false];

    const testCases = labels.flatMap((label) =>
      booleans.flatMap((checked) =>
        booleans.map((disabled) => ({
          label,
          checked,
          disabled,
        }))
      )
    );

    test.each(testCases)(
      "renders correctly with label=%s, checked=%s, disabled=%s",
      ({ label, checked, disabled }) => {
        const { asFragment } = render(
          <Checkbox label={label} checked={checked} disabled={disabled} />
        );

        expect(asFragment()).toMatchSnapshot();
      }
    );
  });
});
