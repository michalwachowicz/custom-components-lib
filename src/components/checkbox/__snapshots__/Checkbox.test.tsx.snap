// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> snapshots renders correctly with label={ label: 'Test', checked: false, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        class="checkbox-input"
        id="«rg»"
        type="checkbox"
      />
    </div>
    <label
      class="checkbox-label"
      for="«rg»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Checkbox /> snapshots renders correctly with label={ label: 'Test', checked: false, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        class="checkbox-input"
        disabled=""
        id="«rf»"
        type="checkbox"
      />
    </div>
    <label
      class="checkbox-label"
      for="«rf»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Checkbox /> snapshots renders correctly with label={ label: 'Test', checked: true, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        checked=""
        class="checkbox-input"
        id="«re»"
        type="checkbox"
      />
    </div>
    <label
      class="checkbox-label"
      for="«re»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Checkbox /> snapshots renders correctly with label={ label: 'Test', checked: true, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        checked=""
        class="checkbox-input"
        disabled=""
        id="«rd»"
        type="checkbox"
      />
    </div>
    <label
      class="checkbox-label"
      for="«rd»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Checkbox /> snapshots renders correctly with label={ label: undefined, checked: false, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        class="checkbox-input"
        id="«rc»"
        type="checkbox"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Checkbox /> snapshots renders correctly with label={ label: undefined, checked: false, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        class="checkbox-input"
        disabled=""
        id="«rb»"
        type="checkbox"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Checkbox /> snapshots renders correctly with label={ label: undefined, checked: true, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        checked=""
        class="checkbox-input"
        id="«ra»"
        type="checkbox"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Checkbox /> snapshots renders correctly with label={ label: undefined, checked: true, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="checkbox-container"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        checked=""
        class="checkbox-input"
        disabled=""
        id="«r9»"
        type="checkbox"
      />
    </div>
  </div>
</DocumentFragment>
`;
